Realm/ObjectStore/impl/async_query.cpp

```diff
@@ -0,0 +1,265 @@
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "async_query.hpp"
+
+#include "realm_coordinator.hpp"
+#include "results.hpp"
+
+using namespace realm;
+using namespace realm::_impl;
+
+AsyncQuery::AsyncQuery(Results& target)
+: m_target_results(&target)
+, m_realm(target.get_realm().shared_from_this())
+, m_sort(target.get_sort())
+, m_sg_version(Realm::Internal::get_shared_group(*m_realm).get_version_of_current_transaction())
+{
+    Query q = target.get_query();
+    m_query_handover = Realm::Internal::get_shared_group(*m_realm).export_for_handover(q, MutableSourcePayload::Move);
+}
+
+AsyncQuery::~AsyncQuery()
+{
+    std::lock_guard<std::mutex> lock(m_target_mutex);
+    m_realm = nullptr;
+}
+
+size_t AsyncQuery::add_callback(std::function<void (std::exception_ptr)> callback)
+{
+    m_realm->verify_thread();
+
+    auto next_token = [=] {
+        size_t token = 0;
+        for (auto& callback : m_callbacks) {
+            if (token <= callback.token) {
+                token = callback.token + 1;
+            }
+        }
+        return token;
+    };
+
+    std::lock_guard<std::mutex> lock(m_callback_mutex);
+    auto token = next_token();
+    m_callbacks.push_back({std::move(callback), token, -1ULL});
+    if (m_callback_index == npos) { // Don't need to wake up if we're already sending notifications
+        Realm::Internal::get_coordinator(*m_realm).send_commit_notifications();
+        m_have_callbacks = true;
+    }
+    return token;
+}
+
+void AsyncQuery::remove_callback(size_t token)
+{
+    Callback old;
+    {
+        std::lock_guard<std::mutex> lock(m_callback_mutex);
+        REALM_ASSERT(m_error || m_callbacks.size() > 0);
+
+        auto it = find_if(begin(m_callbacks), end(m_callbacks),
+                          [=](const auto& c) { return c.token == token; });
+        // We should only fail to find the callback if it was removed due to an error
+        REALM_ASSERT(m_error || it != end(m_callbacks));
+        if (it == end(m_callbacks)) {
+            return;
+        }
+
+        size_t idx = distance(begin(m_callbacks), it);
+        if (m_callback_index != npos && m_callback_index >= idx) {
+            --m_callback_index;
+        }
+
+        old = std::move(*it);
+        m_callbacks.erase(it);
+    }
+}
+
+void AsyncQuery::unregister() noexcept
+{
+    std::lock_guard<std::mutex> lock(m_target_mutex);
+    m_target_results = nullptr;
+    m_realm = nullptr;
+    m_unregistered = true;
+}
+
+void AsyncQuery::release_query() noexcept
+{
+    {
+        std::lock_guard<std::mutex> lock(m_target_mutex);
+        REALM_ASSERT(m_unregistered);
+    }
+
+    m_query = nullptr;
+}
+
+bool AsyncQuery::is_alive() const noexcept
+{
+    std::lock_guard<std::mutex> lock(m_target_mutex);
+    return !m_unregistered;
+}
+
+void AsyncQuery::run()
+{
+    REALM_ASSERT(m_sg);
+
+    // This function must not touch any members touched in deliver(), as they
```

Comment:
I'm a little confused by this comment as there are a number of members accessed within this function that are touched in `deliver()` (e.g., `m_target_results`, `m_have_callbacks`, `m_initial_run_complete`). Can you clarify what this comment is trying to say?